### DISCUSSION TITLE
How to download an output file from a Python notebook (within the Redivis notebook environment)?

### DISCUSSION THREAD

**Discussion by hluling** [redivis/meta#193] [Q&A] [answered]
In a Python notebook (within the Redivis notebook environment), I tried to save a processed dataframe as csv to local drive.

I tried:
`from IPython.display import FileLink`
`df_thin.to_csv("results.csv", index=False)`
`display(FileLink("results.csv"))`

> [results.csv](someurl)

Clicking the above link didn't trigger anything. Then I copied and opened the link, and the browser returned this:

> {"message":"This action is not allowed","status":400}

Is there a way to download such a csv file? I'm asking because there may also be a need to save model run results to local drive.

I think a more general question is how to access the file system in a Redivis notebook environment.

**Comment by imathews** [comment, marked as answer]
Hi @hluling — my apologies for the very delayed reply here, we somehow missed this notification! In the current (v1) implementation, notebooks are designed as a terminus in a project; e.g,. for final graphs and figures. However, we've received a lot of feedback that it would be helpful to be able to export notebook outputs (both via download as well as the ability to materialize dataframes as new tables in the project); we're actively working towards this goal.

The errors that you're seeing are due to security considerations; for high risk datasets that have export restrictions, we need to prevent un-approved data downloads from notebooks. However, it would be much better to only apply these restrictions when the notebook actually contains such restricted data. Let me see if we can implement a quick update that will support your immediate use case (I'll follow up here), and hopefully in the coming months we'll be able to further improve the workflow for accessing the notebook filesystem and moving files + data out of notebooks.

**Comment by anmol-eze** [comment]
have approved CheXpert Plus data access and can manually download individual PNG files. Local API export returns 403 access_denied: This table cannot be exported to the requested environment. I have a curated list of 11,585 specific frontal PNG paths. Please advise the approved method to bulk-export only this filtered subset, or enable export approval for a derived file-index table.

**Reply by imathews** [reply]
Hi @anmol-eze – thanks for reporting this issue. Could you share the URL of the table in question, and we can investigate further what might be going on here.

cc @ashirts
